### DISCUSSION TITLE
Using tag prefix’s in a monorepo

### DISCUSSION THREAD

**Discussion by sheldonhull** [miniscruff/changie#502] [Q&A] [answered]
Is it possible to use components to generate unique component tags that version independently in each directory? I have 5-6 projects in a directory. I want to use changie to version but each project in the repo should be setting a prefixed tag output so I can move those independently. 

The version tag should increment for each independently. 
A tool I used in the past was `svu` that supported this.

I didn’t see any support for pulling the specific semver using something like `changie latest —component ui` would result in `ui/v1.0.2`.

**Comment by miniscruff** [comment]
Components are simply a header in a version, so there would be no way to separate them from the versions they are nested in. Changie latest works by searching the changes directory, so for anything like what you are suggesting to work it would need to either look in a specific directory or some how filter files.

Is your goal to make one shared changelog for all directories? Do you have any example repos or maybe a layout I can review? I have not thought much about how to use changie in monorepos where the directories are versioned separately.

One issue might be that prefixed tags are not proper semver and so there might be issues there.

**Reply by sheldonhull** [reply]
Different changelog for different components is fine. 

prefixes are common on monorepos like kubernetes for example, or some Pulumi repos.  The prefix is for filtering and the semver is still valid after that.

I want specific components to be be independently versioned. Make sense or need more examples.  It’s fine if it doesn’t fit this project, but wanted to start the dialogue.

**Reply by miniscruff** [reply]
I was actually going to start using a monorepo for a side project so I am considering how it would work in changie and how you might want to export or control the generated changelog. After reading your comment and thinking a little I actually think this would be a good feature, just need to find out how. 

( I have rewritten the below content at least three times now... so don't take it too seriously yet )

My current thought process is a new config called something like "projects", name not set in stone of course. Your configuration would look something like:

```yaml
projects:
# label is the label used in prompts for our selection, defaults to key if none specified
- label: UI
  # key is used for file paths and version outputs
  key: ui
  # the changelog value here overrides the global value on a per project basis
  changelog: ui/CHANGELOG.md
- label: Email Sender
  key: email_sender
  changelog: services/email/CHANGELOG.md
projects_version_separator: "/"
```

This configuration gives us the ability to create a changelog like [stellar](https://github.com/stellar/go/blob/master/CHANGELOG.md) a fairly large monorepo with split changelogs. Internally the unreleased changes remain in the changes directory and release notes can also live there under the key per project. Like so

```
your_app/
├─ .changes/
│  ├─ unreleased/
│  │  ├─ whatever.yaml
│  ├─ email_sender/
│  │  ├─ v0.1.0.md
│  │  ├─ v0.2.0.md
│  ├─ ui/
│  │  ├─ v0.1.0.md
```

This would affect a few commands like you already mentioned with latest.
```shell
# new will stay the same and the first prompt will be the project
changie new
# but you can also include the project like with other values
changie new --project ui
# batch requires you to provide the project you are batching for
changie batch auto --project ui
changie batch auto # fails if projects are enabled, maybe in the future we can figure it out
# merge one project at a time
changie merge --project ui
# merging all projects is probably a safe action as well
changie merge
# latest and next both require a project
changie latest --project ui
changie next --project ui
```

The output of latest and next would be the following formula:
project key + project_version_separator + latest/next version
For example given the configuration above and versions in the example above the outputs would be:
```sh
$ changie latest --project email_sender
email_sender/v0.2.0
$ changie next patch --project ui
ui/v0.1.1
```

Let me know your thoughts or if you have any questions on the design.

**Reply by marcosnils** [reply]
Looking good as an initial release. The only downside I see is that it's not easily plugable for users that currently have multiple change files per folder and multiple `.changes` folder.

**Reply by miniscruff** [reply]
Fair, it would need manual edits to at least a few files but I think it's mostly just moving them around. I could write up a guide for the steps of moving from N to 1 config and changes directory. The goal is that the output is the same just in different places without the need for anyone to cd to manage fragments or releases.

**Reply by miniscruff** [reply]
I think it would be best to develop and release this feature in a preview mode until any issues can be worked out, I have not done that for Changie before as all new features have always required a config change anyway. I do want to make sure the config changes and flow work well and feel good which I will have a hard time testing myself. 

A base idea I have for that is simply displaying a warning prompt when using `changie new` and `changie batch` if projects are configured that they are in "preview mode" and mention that in the release notes until it is stable and good to go. 

Other then that, its just a timing thing to get the work done.

**Reply by marcosnils** [reply]
> I think it would be best to develop and release this feature in a preview mode until any issues can be worked out, I have not done that for Changie before as all new features have always required a config change anyway. I do want to make sure the config changes and flow work well and feel good which I will have a hard time testing myself.

Yep, sounds like a good plan.

> A base idea I have for that is simply displaying a warning prompt when using changie new and changie batch if projects are configured that they are in "preview mode" and mention that in the release notes until it is stable and good to go.

Also SGTM!

**Reply by miniscruff** [reply, marked as answer]
Created an issue from my design above, no ETA on timing but it is at least visible in an issue.
https://github.com/miniscruff/changie/issues/512

Edit: Release in v1.13.0 https://github.com/miniscruff/changie/releases/tag/v1.13.0

**Comment by gerhard** [comment]
We were talking about this today with @grouville (same context as @marcosnils - https://github.com/miniscruff/changie/discussions/506#discussion-5422854) and either solution seems simple enough to me:
1. Add the change as a single command, e.g. `changie new --kind "Fixed" --body "Show when Engine is loading in the output" --custom "Author=TomChv" --custom "PR=5488"`, then re-run the command in all directories that need it 💡 FWIW, I am a big fan of [`fc`](https://en.wikipedia.org/wiki/Fc_(Unix)).
2. Create the change once, then just `cp` the resulting YAML file in all directories that need it, e.g. `cp {,sdk/go/}.changes/unreleased/Fixed-20230715-131052.yaml`

WDYT @marcosnils, would either of the above be enough in a monorepo?

**Reply by marcosnils** [reply]
> WDYT @marcosnils, would either of the above be enough in a monorepo?

Yes, IMO either is enough given the current state of changie. Having said that, what I'd really like is if this "projects" idea move forwards is to add `--projects` to changie new and let changie handle the it internally :pray:

**Comment by sheldonhull** [comment]
Implementation so far seems great!

Here's some feedback @miniscruff.

- If changie new is changie and a project is required, make it a selector.
- `changie latest --project myproject --remove-prefix` removes the v prefix. If I get the latest version for a project I'd actually expect the semver version without the project name. Not sure best way to do this, if `--remove-project` from the version makes sense or if the default behavior would be to remove the prefix. That's sort of what I'd expect.
- I think to do tagging I need `changie changesdir --project myproject` to return the location of the pending changes. I use `git tag -a "$(changie latest --project 'myproj')" -F ".changes/$(changie latest --project 'myproj').md"` and now if I move or rename things would break.
- Idea: I have to script around the tagging, so having the equivalent of `changie tag --project myproj` and having a new git tag added with the message being the changelog file like my snippet would be a big help. I can keep scripting that, but since it's a versioning tool having it handle the tag addition and check for it existing would cut out more boilerplate. (For another example of this see SVU tool https://github.com/caarlos0/svu which I've used successfully in the past).

**Reply by miniscruff** [reply]
Thanks for the feedback, here are my responses.

1. Not sure what you mean by selector here, could you explain this more?

2. I think adding `--remove-project` makes sense. The --remove-prefix would not remove the project but could be combined with remove project to remove both. I also think its easier to include the project with the option to remove it, rather then having the user try and rebuild the full version by hand after the fact. 

3. With the above change your git command would be
```sh
PROJECT=my_proj
git tag \
    -a "$(changie latest --project $PROJECT)" \
    -F ".changes/$PROJECT/$(changie latest --project $PROJECT --remove-project).md"
```
The path to the latest release for a project is changes_dir config + project key + latest version, so that should be fine. This command is quite long and complex, but is like that because using projects is an additional layer of complexity. This command is a bit easier without projects of course.

4. Changie has no concept of source control, it only reads and writes files. I can add a small page on how to do this work to the docs but changie itself will never read or write to source control. That is one of the design principles it is built from. SVU itself is a git based tool, Changie is not.

**Reply by sheldonhull** [reply]
Selector… when you pick components in changie new you can select from list when entering changie new. 

I’d like the user to not have to remember the key in a repo and be able to select. 


good responses! Cheers

**Reply by miniscruff** [reply]
Oh yea, project is the first prompt required if they are enabled.

**Reply by sheldonhull** [reply]
Ok so if I had project enabled and did changie new and it errored since I didn’t provide project that’s a bug?

**Reply by miniscruff** [reply]
Uhh yea that sounds bad lol. Please do post that. These are the concerns I had and why I marked it as a preview.

**Reply by sheldonhull** [reply]
Non-issue. I needed to run `aqua generate -i miniscruff/changie` to bump to version 1.13. For some reason maybe I was still running one edition prior.
